paasta_tools/paasta_metastatus.py

```diff
@@ -147,13 +169,68 @@ def utilization_table_by_grouping_from_mesos_state(
             [v for g, v in grouping_values],
             healthcheck_utilization_pairs,
-        ) + [str(resource_info_dict['slave_count'])])
+        ))
+        # Fill table rows with service-instance data if possible.
+        if service_instance_stats:
+            fill_table_rows_with_service_instance_stats(service_instance_stats, resource_utilizations, table_rows)
+
+        # Always append the agent count last
+        table_rows[-1].append(str(resource_info_dict['slave_count']))
+
     table_rows = sorted(table_rows, key=lambda x: x[0:len(groupings)])
     all_rows.extend(table_rows)
 
     return all_rows, healthy_exit
 
 
+def fill_table_rows_with_service_instance_stats(
+    service_instance_stats: Dict[str, float],
+    resource_utilizations: List[metastatus_lib.ResourceUtilization],
+    table_rows: List[List[str]],
+):
+    # Calculate the max number of runnable service instances given the current resources (e.g. cpus, mem, disk)
+    resource_free_dict = {rsrc.metric: rsrc.free for rsrc in resource_utilizations}
+    num_service_instances_allowed = float('inf')
+    limiting_factor = 'Unknown'
+    # service_instance_stats.keys() should be a subset of resource_free_dict
+    for rsrc_name, rsrc_amt_wanted in service_instance_stats.items():
+        if rsrc_amt_wanted > 0:
+            # default=0 to indicate there is none of that resource
+            rsrc_free = resource_free_dict.get(rsrc_name, 0)
+            if rsrc_free // rsrc_amt_wanted < num_service_instances_allowed:
+                limiting_factor = rsrc_name
+                num_service_instances_allowed = rsrc_free // rsrc_amt_wanted
+    table_rows[-1].append('{:6} ; {}'.format(int(num_service_instances_allowed), limiting_factor))
+
+
+def get_service_instance_stats(service: str, instance: str, cluster: str) -> Dict[str, float]:
+    """Returns a Dict with stats about a given service instance.
+
+    Args:
+        service: the service name
+        instance: the instance name
+        cluster: the cluster name where the service instance will be searched for
+
+    Returns:
+        A Dict mapping resource name to the amount of that resource the particular service instance consumes.
+    """
+    if service is None or instance is None or cluster is None:
+        return {}
+
+    service_instance_stats: Dict[str, float] = {}
+    try:
+        instance_config = get_instance_config(service, instance, cluster)
+        # Get all fields that are showed in the 'paasta metastatus -vvv' command
+        service_instance_stats['mem'] = instance_config.get_mem()
+        service_instance_stats['cpus'] = instance_config.get_cpus()
+        service_instance_stats['disk'] = instance_config.get_disk()
+        service_instance_stats['gpus'] = instance_config.get_gpus()
+    except Exception as e:
+        log.error(f'Failed to get stats for service {service} instance {instance}')
+        log.error(repr(e))
+    return service_instance_stats
```

Comment:
It it OK to return potentially partially constructed `service_instance_stats ` after an exception?

Reply:
It should return an empty dict so I've made it more explicit

Reply:
After using TypedDict it has to return None since it can't return an empty dict